### ISSUE TITLE
Map this accessor should return `const ref`/`ref` for non-nilable classes

### ISSUE THREAD

**Issue by bmcdonald3** [chapel-lang/chapel#21949]
Due to misunderstanding https://github.com/chapel-lang/chapel/issues/21395, I had implemented the map `proc this` to always return a borrow, but @mppf raised that it should be returning a reference to a value of class type rather than a different type.

The below program will print `borrowed C`, when we would like that to print `shared C`.

```chpl
use Map;

class C { }

var m: map(int, shared C);

m.add(1, new shared C());

writeln(m[1].type:string);
```

Compare with a `list` which prints `shared C` for the type the accessor returns:

```chpl
use List;

var ll: list(shared C);
ll.append(new shared C());

writeln(ll[0].type:string);
```

**Comment by bradcray** [comment]
Is this self-evident?  It doesn't particularly surprise me that it would return a borrowed, but I can see an argument either way.  What does/should an access of a map of borrowed C return?

**Comment by bradcray** [comment]
> > What does/should an access of a map of borrowed C return?
> 
> It returns a borrow but I think that is unsurprising.

Oops, I typo'd my question.  I meant to ask about a map of `unmanaged`.  Based on your answer, I'm imagining it would return a `const ref` or `ref` to the `unmanaged`?

And I'm also imagining the same would be true for `owned`?  (i.e., `const ref` or `ref` to the `owned`).  For some reason, in my reading of the OP, I was thinking that the proposal was that `owned` should return a `borrow` to the class but that the others should return a copy-out of the class.

(and I'm fine with the proposal assuming I'm on the right track now).

**Comment by mppf** [comment]
> Oops, I typo'd my question. I meant to ask about a map of `unmanaged`. Based on your answer, I'm imagining it would return a `const ref` or `ref` to the `unmanaged`?

That's what I would expect.



> And I'm also imagining the same would be true for `owned`? (i.e., `const ref` or `ref` to the `owned`). For some reason, in my reading of the OP, I was thinking that the proposal was that `owned` should return a `borrow` to the class but that the others should return a copy-out of the class.

That's also what I would expect, but Ben's original wording of this issue described a different behavior for `owned` vs `shared` in this regard. Since it was summarizing my concern, I just edited it (so it no longer says that).

**Comment by bradcray** [comment]
Thanks.  And then focusing on the title of this issue, is it the case that `main` currently returns a ref/const ref for all class types other than `shared` today?  Or should the title be updated to cover all classes as well?

**Comment by mppf** [comment]
> Thanks. And then focusing on the title of this issue, is it the case that `main` currently returns a ref/const ref for all class types other than `shared` today? Or should the title be updated to cover all classes as well?

Looking at the code, it looks like it is treating non-nilable classes specially (and returning a borrow of these).

I updated the title.

**Comment by bmcdonald3** [comment]
Since we would always be returning a `ref`/`const ref` here, would we be expecting users to put an explicit `.borrow()` on the map access for uses that would transfer ownership?

So, something like this:
```
use Map;

class C {}

var m = new map(int, owned C);

m[1] = new owned C();

const a = m[1].borrow(); // user explicitly borrows
```

Since `m[1]` would be of type `owned C`, we can't do just the regular `const a = m[1]`. Am I understanding this correctly? (this is consistent with the behavior of `list`, but just wanted to make sure I had things straight this time before making a bunch of code changes)

**Comment by bmcdonald3** [comment]
I've opened up https://github.com/chapel-lang/chapel/pull/22051 for this and feel much more confident in this implementation than the previous one (the reason that I had difficulty porting to Arkouda was that the behavior I had put in before made it so we couldn't get the references out of a map and I've confirmed that this lets us clean up the Arkouda code that I added to work around that).

The new behavior in the PR for `map.this` always returns a `ref` or `const ref` and throws an error when a non-default initializable class is attempted to be accessed by `this` (since we don't know how to fill the slot to get a reference to without default initializing the value), but will default initialize a value and place it in the slot, returning a reference to that value otherwise. I believe this is inline with `list` and matches what has been discussed in this issue, but let me know if anything about that sounds odd.
